### ISSUE TITLE
AOT Error resolving funcion symbol statically. Worked in 2.3.1 and not in 2.4.0

### ISSUE THREAD

**Issue by emilio-martinez** [angular/angular#13605]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I have my own property decorators (used for validation) applied throughout a sizable Angular project. These decorators have had no issues with `ngc` with the exception of maybe requiring one of the params passed to the decoration to be exported. However, as of `2.4.0` I am getting the following error:

```
Error: Error encountered resolving symbol values statically. Calling function '[function name]',
function calls are not supported. Consider replacing the function or lambda with a reference to
an exported function.
```

To provide more detail, the decorator allows to be passed a custom validation function as part of one of it's parameters. I attempted to save the passed function value as an exported constant and still received the same error.

Reverting to `2.3.1` "resolves" the issue. Of course there are ways to work around this, but I thought I should bring it up in case there's something that was not caught in regression testing or something of the sort.

**Expected behavior**
AOT compilation with no error. I have been able to compile with these decorators since around `rc.6` days, around the time that I wrote them.

**Minimal reproduction of the problem with instructions**
Sorry, cannot reproduce since it's an AOT compilation issue. If it's helpful, this would be a simplified representation of what I described above:
```ts
export const validationOptions = { option: optionValue };
@MyValidationDecorator( TypeEnum.type, ValidationFunction(validationOptions) )
myClassProp: type;
```

Additionally, also as mentioned above, the following doesn't work either, attempting to follow the indication of "Consider replacing the function or lambda with a reference to an exported function.":
```ts
export const validation = ValidationFunction({ option: optionValue });
@MyValidationDecorator( TypeEnum.type, ValidationFunction(validation) )
myClassProp: type;
```

**What is the motivation / use case for changing the behavior?**
Ideally, being able to move forward with the same path I had been for a few months now. Like I said, there's ways to work around this, but ideally I wouldn't have to unless theres a specific Angular Compiler need/benefit to this process. I looked at the change log and there are some compiler changes around metadata, but I did not see something that immediately jumped at me as the cause of this Error.

**Please tell us about your environment:**
Mac OS X 10.11.6, VSCode 1.9.0-insider, npm 3.10.8, webpack 2.2.0-rc.1, webpack-dev-server 2.2.0-rc.0, Typescript 2.0.10

* **Angular version:** 2.4.0. Not an issue with 2.3.1

* **Browser:** Not in browser
 
* **Language:** Typescript 2.0.10

* **Node (for AoT issues):** `node --version` = `v6.9.1`

**Comment by slubowsky** [comment]
Getting same thing with immutable js Record.

```
import { Record } from 'immutable';

export const PersonRecord = Record({
    id: '',
    name: ''
});
export class Person extends PersonRecord {
    id: number;
    name: string;

    constructor(props: any) {
        super(props);
    }
}
```

When compiled (angular cli - ng serve --aot)
```
ERROR in Error encountered resolving symbol values statically. Calling function 'Record', function calls are not supported. Consider replacing the function or lambda with a ref
erence to an exported function, resolving symbol PersonRecord in C:/dev/Person-APP/src/app/shared/person.ts, resolving symbol Person in C:/dev/Person-APP/src/app/shared/person.ts
```

Trying
```
export function createRecord() {
    return Record({
        id: '',
        name: ''
    });
}
export const PersonRecord = createRecord();
```
Didnt help.

**Comment by DzmitryShylovich** [comment]
@slubowsky It's a known issue https://github.com/angular/angular/issues/13436

**Comment by slubowsky** [comment]
@DzmitryShylovich Thanks. Just moved from 2.2.4 to 2.4 with new version of cli today and didn't see that issue, it's a few pages back by now. Definitely a regression though. I've been doing this  forever (not always using aot though but occasionally)

**Comment by mischkl** [comment]
Possibly related as it also broke starting with 2.4: https://github.com/angular/angular/issues/13614

**Comment by DzmitryShylovich** [comment]
https://github.com/rangle/angular-2-aot-sandbox

**Comment by emilio-martinez** [comment]
@DzmitryShylovich is there any specific change that would break my use case with `ngc` from `2.3.x` to `2.4.x`? I understand the reasons why the code has to be statically analyzable; however, this used to work since the `~rc.5` days up to `2.3.1`, which is why I was wondering if this is a regression issue. Other than that, there's ways I know I can architect around this, but mostly I want to understand if there's any new practice that I should be following before embarking on any changes.

I looked through the cases listed on https://github.com/rangle/angular-2-aot-sandbox (thanks for the reference, by the way!), and none of those are really match this one. As I'm sure many have, I've run into several of those along the way, and most of them have rarely been a problem to address.  My currently breaking use case, however, is around my own decorator and validation function, all of which I should mention stand on their own—they do not depend on any libraries or frameworks other that being written in Typescript.

**Comment by chuckjaz** [comment]
@emilio-martinez I will attempt to come up with a reproducible case given your description but can you also to produce a minimal example and create a github repository that I can clone with the repro?

**Comment by chuckjaz** [comment]
@emilio-martinez I believe I have a test case and no longer need the repository I requested above.

I believe the issue stems from a refactoring we did in commit 33910ddfc98d092e514d90a2cec3b5daaf3d4277 where we refactored the static reflector to simplify how it handles calls. Unfortunately, this causes us now to try to evaluate the parameters to all decorators, not just the angular decorator.

I am working on a fix now.
